modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/services/VaultWatcherService.kt

```diff
@@ -30,35 +32,43 @@ val UNLOCKER: ScheduledExecutorService = Executors.newSingleThreadScheduledExecu
 const val PLACE_HOLDER: String = "THIS_IS_A_PLACE_HOLDER"
 
 @CordaService
-class VaultWatcherService(tokenObserver: TokenObserver? = null) : SingletonSerializeAsToken() {
-
+class VaultWatcherService(private val tokenObserver: TokenObserver, private val serviceHub: ServiceHub) : SingletonSerializeAsToken() {
     enum class IndexingType {
-        EXTERNAL_ID, PUBLIC_KEY
+        EXTERNAL_ID, // external id (all keys registered to this id as an owner) + token class + token identifier
+        PUBLIC_KEY, // Public key + token class + token identifier
+        TOKEN // Just token class + token identifier without owner
     }
 
     private val cache: ConcurrentMap<TokenIndex, TokenBucket> = ConcurrentHashMap()
     // TODO for now not used, add some reasonable default - stop listening print massive error; Figure out sensible default
     //    5% of heap, your heap needs to be min 8GB?
     private var cacheSize: Int = 1024
 
-    constructor(appServiceHub: AppServiceHub) : this(getObservableFromAppServiceHub(appServiceHub))
+    constructor(appServiceHub: AppServiceHub) : this(getObservableFromAppServiceHub(appServiceHub, null), appServiceHub)
 
     companion object {
         val LOG = contextLogger()
 
-        private fun getObservableFromAppServiceHub(appServiceHub: AppServiceHub): TokenObserver {
+        private fun getObservableFromAppServiceHub(appServiceHub: AppServiceHub, providedConfig: InMemorySelectionConfig?): TokenObserver {
             val config = appServiceHub.cordappProvider.getAppContext().config
-            val configOptions = InMemorySelectionConfig.parse(config)
-            // TODO cacheSize = configOptions.cacheSize
-
-            val ownerProvider: ((StateAndRef<FungibleToken>, AppServiceHub) -> Any) = if (configOptions.indexingStrategy == IndexingType.PUBLIC_KEY) {
-                { stateAndRef, _ ->
-                    stateAndRef.state.data.holder.owningKey
-                }
-            } else {
-                { _, _ ->
-                    throw IllegalStateException("Only IndexingType.PUBLIC_KEY available on Corda V4")
-                }
+            val configOptions = if (providedConfig == null) {
+                InMemorySelectionConfig.parse(config)
+            } else providedConfig
+
+            val ownerProvider: ((StateAndRef<FungibleToken>, ServiceHub) -> Holder) = when (configOptions.indexingStrategy) {
```

Comment:
This is a bit "special" :) Perhaps we can tidy up at some point. Not essential for now though.

Reply:
What is special about it? This is function on service construction, you can pass indexing only via config options (services have access only to `AppServiceHub`). To be able to index by anything you would have to have some translation layer between config and owner provider.